Add VDI profiles based on lpagg

uvchik opened the pull request:
The basic code is taken from https://github.com/jnettels/lpagg but most of the code has been revised for modular use and to speed up the code. The lpagg package contains various functionality but only the VDI-profiles and reading the TRY data from DWD is used.

To test the new profiles use the `vdi_profile_example` in the examples section. The parameter of the houses are added within a loop. I used this to multiply the number of houses to check how long it would take to model e.g. 1000 houses. 1000 houses on an hourly base will take less than a minute, which shows that the draft is at least fast enough :racing_car: 

Related to #36

Install this branch to test the code:
```
pip install https://github.com/oemof/demandlib/archive/features/add-vdi-from-lpagg
```

It is the first draft, I am open for concrete ideas.

 * [ ] add tests
 * [ ] add documentation
 * [ ] clean code

uvchik committed:
Add new methods

uvchik committed:
Add usage for electrical profiles in the docs

uvchik committed:
Ignore long line with url in pep checker

uvchik committed:
Fix pandas FutureWarning

uvchik committed:
Revise methods of electrical profile to make usage clearer

uvchik committed:
Update tox checks using twine

uvchik committed:
Add tests for tools and elecrical profiles

uvchik committed:
Fix isort issue

uvchik committed:
Add holidays as list

uvchik committed:
Make attributes protected

uvchik committed:
Update seasons instead of replacing them

uvchik committed:
Add test for changed season periods

uvchik committed:
Add warning in docstring

uvchik committed:
Change equation to latex and add axis

uvchik committed:
Add sources to docs

uvchik committed:
Merge pull request #45 from oemof/features/revise-methods-for-elec-slp

Features/revise methods for elec slp

uvchik committed:
Merge branch 'dev' of github.com:oemof/demandlib into v0.2dev

uvchik committed:
Add lpagg code to demanlib

uvchik committed:
Add lpagg code to demanlib

uvchik committed:
Add example for VDI profiles

uvchik committed:
Add example for VDI profiles

uvchik committed:
Fix grammar

uvchik committed:
Add extra requirements for examples

uvchik committed:
Use black style

uvchik committed:
Use black style

uvchik committed:
Revise region module

uvchik committed:
Revise region module

uvchik committed:
Make it possible to import Region from demandlib.vdi

uvchik committed:
Make it possible to import Region from demandlib.vdi

uvchik committed:
Add function to find TRY region by coordinates

uvchik committed:
Add function to find TRY region by coordinates

uvchik committed:
Add geometry as extra requirements

uvchik committed:
Add docstrings

uvchik committed:
Add docstrings

uvchik committed:
Remove binary file from repository

uvchik committed:
Remove binary file from repository

uvchik committed:
Do not read unused columns

uvchik committed:
Do not read unused columns

uvchik committed:
Fix style issues

uvchik committed:
Fix style issues

uvchik committed:
Fix style issues

uvchik committed:
Fix style issues

uvchik committed:
Use new black style

jnettels commented:
@uvchik thanks very much, especially the speed looks promising! Not iterating through rows does indeed help a lot :-)
I was able to run your example. I have reduced it to a calculation for a single house, and I am still getting differences between my implementation and yours, even if they are very small. It comes down to me identifying e.g. 21. October 2017 in TRY04 as WWB, while your code yields WWH. The difference is the daily mean of the cloud cover (< or >5). This only happens for 4 days, I think.

I am not 100% sure yet, but I think it comes down to the following: At 15min frequency, Pandas df.resample("D").mean() will use the values between (including) ``2017-10-21 00:00`` and ``2017-10-21 23:15`` to give the mean of ``2017-10-21``. However, in the weather data, ``2017-10-21 00:00`` describes the weather conditions between ``2017-10-20 23:00`` and ``2017-10-21 00:00``. The value in ``2017-10-21 00:00`` therefore belongs to ``2017-10-20``. To put it more simply: By DWD definition, the weather data does not start at ``2017-1-1 00:00``, but ``2017-1-1 01:00``.
You seem to ignore that and simply load the weather data, starting at ``2017-1-1 00:00``, which may actually solve those problems. It may very well be that your solution is correct, and my many workarounds lead to incorrect results. Whenever I think about this stuff, I have to double and tripple check.

Some more stuff that might need consideration:
- You seem to allow leap years, but that would lead to missing rows in the weather data. Setting ``vdi.Region(year=2020)`` currently throws ``ValueError: Length mismatch: Expected 8760 rows, received array of length 8784`` and a more helpful message might be appropriate
- The defining temperature for summer should be exposed to the user, with default 15°C (``Tamb_heat_limit`` in my code)
- If no weather data is given, the default weather file for the given TRY region should be loaded
- The user should be able to override the used weather timeseries. (These could then be used for lead years, if desired.)
- While using ``vdi.Region(holidays=holidays.country_holidays('DE', subdiv='NI'))`` from https://pypi.org/project/holidays/ already works (yay!), it could be more encouraged in a final example. I think that project is awesome.

There might be more stuff, but these are all minor issues I think. I need some time for more investigation, though, to see if the API fits all the needs that I can come up with.

uvchik committed:
Use temperature limits from config file

uvchik committed:
Use temperature limits from config file

uvchik commented:
* [x] The weather data seems to be fine to me now. The DWD data starts with 01-01-01 (Month-Day-Hour), while weather data is typically right stamped. So in my left stamped index the first value is 2017-01-01 00:00. For me a right stamped index seems to be counter intuitive for a daily index
* [ ] My idea is to duplicate the nearest day with the same weekday to create the typical day for the 29th of February. If it is a Sunday I would use the last Sunday, if it is a weekday I would use the last weekday which would be the 28th or the 27th. I think this is good enough regarding all the other assumptions. (still a ToDo)
* [ ] I am not sure whether we should offer this or not. Some of the factors are connected to the TRY region. I seems to be strange to use an arbitrary data set with this procedure. In that case one has to pass the weather data and the TRY region.
* [ ] I use [workalendar](https://github.com/workalendar/workalendar) in my example, but feel free to use your favourite holiday package.

uvchik committed:
Add missing config import

uvchik committed:
Add missing config import

uvchik committed:
Fix time stamp to left index

uvchik committed:
Fix time stamp to left index

uvchik committed:
Fix filepath for TRY data

uvchik committed:
Fix filepath for TRY data

pep8speaks commented:
Hello @uvchik! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









* In the file [`src/demandlib/config.py`](https://github.com/oemof/demandlib/blob/2e205c96dbcd996629ce5cfe64b7b86da34c58f7/src/demandlib/config.py):

> [Line 44:5](https://github.com/oemof/demandlib/blob/2e205c96dbcd996629ce5cfe64b7b86da34c58f7/src/demandlib/config.py#L44): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

uvchik committed:
Remove blank line

uvchik committed:
Remove blank line

jnettels commented:
Here are the notes about what we discussed:
- Make ``summer_temperature_limit`` (currently set in the ini file) an attribute of each house. As a consequence, different series of typical days have to be created for each unique temperature. These typical days then have to be correctly used for the correct houses
- Allow manipulation of the weather data. Using the default weather data should be the default behaviour, but using your own weather data should also be possible. For example, ``vdi.Region`` could have a function ``get_weather()``  which would yield a DataFrame which can be manipulated and a function ``set_weather()`` which overwrites that default weather. This can be used before calling ``get_load_curve_houses()``.

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,559 @@
+                    load_curve_house[column] = (
+                        load_curve_house[column] / sum_ * q_a
+                    )
+            load_curve_house.columns = pd.MultiIndex.from_product(
+                [
+                    [house["house_type"]],
+                    load_curve_house.columns,
+                ]
+            )
+            house_profiles[house["name"]] = load_curve_house
+        return pd.concat(
+            house_profiles.values(), keys=house_profiles.keys(), axis=1
+        )
```
My suggestion for named levels:
```Python
            load_curve_house.columns = pd.MultiIndex.from_product(
                [
                    [house["house_type"]],
                    load_curve_house.columns,
                ],
                names=['house_type', 'energy']
            )
            house_profiles[house["name"]] = load_curve_house
        return pd.concat(
            house_profiles.values(), keys=house_profiles.keys(), axis=1,
            names=['name']
        )
```

p-snft commented on examples/electricity_demand_example.py:
```diff
@@ -27,7 +27,7 @@
 import demandlib.particular_profiles as profiles
 
-# The following dictionary is create by "workalendar"
+# The following dictionary has been created by "workalendar"
```
```suggestion
# The following dictionary is created by "workalendar"
```
I am a bit puzzled here. First, you write that the dictionary has been/ is created using `workalendar` but later you define the holidays yourself.

jnettels replied:
While I am writing comments here... :-)
Since those lines are commented out, I think this is just to suggest how to use an alternative library, without creating a new dependency for the installation of demandlib. My only suggestion would be to also name https://pypi.org/project/holidays/ as another option. Personally, it took me much too long to stumble across it.

p-snft commented on examples/vdi_profile_example.py:
```diff
@@ -0,0 +1,103 @@
+
+from demandlib import vdi
+
+# The following dictionary has been created by "workalendar"
+# pip3 install workalendar
+# from workalendar.europe import Germany
+# cal = Germany()
+# holidays = dict(cal.holidays(2017))
+
+holidays = {
+    datetime.date(2017, 1, 1): "New year",
+    datetime.date(2017, 4, 14): "Good Friday",
+    datetime.date(2017, 4, 17): "Easter Monday",
+    datetime.date(2017, 5, 1): "Labour Day",
+    datetime.date(2017, 5, 25): "Ascension Thursday",
+    datetime.date(2017, 6, 5): "Whit Monday",
+    datetime.date(2017, 10, 3): "Day of German Unity",
+    datetime.date(2017, 10, 31): "Reformation Day",
+    datetime.date(2017, 12, 25): "Christmas Day",
+    datetime.date(2017, 12, 26): "Second Christmas Day",
+}
```
Same as for examples/electricity_demand_example.py.

uvchik replied:
I think it is a good way to let the example run without installing workalendar. For me this has two advantages:

1. Everbody can see how a holiday dict looks like and can adapt it or create a new one
2. The comment explains how it has been created, so that everybody is free to install workalendar and create the dict using it.

p-snft commented on examples/vdi_profile_example.py:
```diff
@@ -0,0 +1,103 @@
+for n in range(1):
+    my_houses.append(
+        {
+            "N_Pers": 3,
+            "name": "EFH_{0}".format(n),
+            "N_WE": 1,
+            "Q_Heiz_a": 6000,
+            "TRY": 4,
+            "copies": 24,
+            "house_type": "EFH",
+            "sigma": 4,
+            "Q_TWW_a": 1500,
+            "W_a": 5250,
```
Language: "EFH", "WE", "Heiz" and "TWW" are German, all others should work in English. (First, I misinterpreted WE as weekend.)

Comprehensibility: What does W_a mean?

jnettels replied:
This has bugged me, too. The reason is sticking to the notation in the norm itself. Even though there is an English version of VDI 4655, that one still uses the German notation. To name a few:
Q: Heat
W: Electrical work
TWW: Domestic hot water
TT: Typical day
a: Year
WE: Flats

Therefore ``W_a`` is the electrical work of one year in kWh.

While this may not be ideal, I'd still argue that sticking to the source notation is very important, at least for the internal code. Making the public API English is something that might be worth considering, as a compromise.

uvchik replied:
You are right, it might be better to use English abbreviations in the code.

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,559 @@
+
+    *VDI-Platz 1, 40468 Duesseldorf, Germany*
+
+Reproduced with the permission of the Verein Deutscher Ingenieure e.V.,
+for non-commercial use only.
```
I fear we may not implement/use this for an open source project. I will stop my review here.

jnettels replied:
Hello there,
in 2019 I asked for permission from the VDI. I explicitly mentioned to the VDI that the implementations of VDI norms would become public code on GitHub as part the work done for scientific projects. I received that permission via email with the only prerequisite being the reference to the VDI norms and the note „Wiedergegeben mit Erlaubnis des Verein Deutscher Ingenieure e. V.“ Afterwards I published my code in "lpagg", which uvchik now adapted for "demandlib". If that helps, I could forward you the full email.

p-snft replied:
If the implementation contains code provided by the VDI, the VDI has to agree to a license. If the code was not provided by the VDI, I doubt that it has copyright. (At least if the code is an implementation and not "just" a derived work.)

If using _the implementation_ is really only allowed for non-commercial users, that is contradicting every possible open source license. (Plus: Non-commercial is not adequately defined. If your institution has third-party funded projects, it can be considered commercial. Just one of the many reasons, why there is no open source just for "non commercial".)

uvchik replied:
As we do have the permission to write public code using the calculations and values of VDI 4655 we can continue.

I remove the "for non-commercial use only" because it is not in line with the used MIT license. Furthermore I remove the copyright in the reference to make sure that it is a reference and not the copyright of the code.

p-snft reviewed (changes requested):
I see copyright issues because of the implementation of the standard:

> Copyright:
>     Verein Deutscher Ingenieure e.V.
>     VDI Standards Department
>     VDI-Platz 1, 40468 Duesseldorf, Germany
> Reproduced with the permission of the Verein Deutscher Ingenieure e.V.,
> for non-commercial use only.

uvchik committed:
Make VDI header clearer

uvchik committed:
Make VDI header clearer

jnettels commented:
I took the liberty to make two changes based on my previous suggestions: Named column levels and the ability to set my own weather file path. These changes help me to use demandlib within lpagg.

I think now the most important remaining issue is the ability to define the ``summer_temperature_limit`` for each house. My other suggestions are just for convenience.

uvchik committed:
Temperature limit is part of the house data

The temperature limit is connected to the efficiency status of the
house.

uvchik committed:
Temperature limit is part of the house data

The temperature limit is connected to the efficiency status of the
house.

jnettels commented:
I have not tested it yet, but thanks for adding the temperature limits. Though, since you removed my solution to define a custom weather file, I need to ask: Was it rude (or just inconvenient) of me to push to this branch, or did you not agree with my approach?

p-snft commented on src/demandlib/vdi/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from .dwd_try import find_try_region  # noqa: F401
+from .regions import Region  # noqa: F401
```
```suggestion
from .dwd_try import find_try_region
from .regions import Region

__all__ == [
    "find_try_region",
    "Region",
]
```

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+
+        Returns
+        -------
+
```
Missing documentation of the return value.

p-snft reviewed (changes requested):
I like the verbose inline documentation explaining what the code does. There is just one location where I found a missing piece of information.

More general, I am still not sure about the usage of German abbreviations. It might or might not be advised. (As the English version of the norm is using them, it can at leats be justified.)

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+        return set(
+            [
+                t_limit(
+                    summer=h["summer_temperature_limit"],
+                    winter=h["winter_temperature_limit"],
```
Do not fail if the temperature options are not defined. Use the default values instead:
```suggestion
                    summer=h.get("summer_temperature_limit", 15),
                    winter=h.get("winter_temperature_limit", 5),
```

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+        for house in self.houses:
+            t_limit = namedtuple("temperature_limit", "summer winter")
+            tl = t_limit(
+                summer=house["summer_temperature_limit"],
+                winter=house["winter_temperature_limit"],
```
```suggestion
                summer=house.get("summer_temperature_limit", 15),
                winter=house.get("winter_temperature_limit", 5),
```

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+        seasons=None,
+        holidays=None,
+        houses=None,
+        resample_rule=None,
```
I am also adding the code snippets for changing the weather file.
```suggestion
        resample_rule=None,
        file_weather=None,
```

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+            dictionary need to have the following keys.
+        resample_rule : str
+            Time interval to resample the profile e.g. 1H (1 hour) or 15min.
+            The value will be passed to the pandas resample method.
```
```suggestion
            The value will be passed to the pandas resample method.
        file_weather : str (optional)
            Path to a 'test reference year' (TRY) weather file by German DWD
            (Deutscher Wetterdienst). If None, the file fitting the given
            try_region will be loaded.
```

p-snft replied:
As long as you adhere to the TRY format, the actual file does not need to be provided by DWD.

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+        self._try_region = try_region
+
+        self._year = year
+        self.weather = None
```
```suggestion
        self.weather = None
        self.file_weather = file_weather
```

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+        days.pop("date")
+
+        # Fetch weather data
+        fn_weather = os.path.join(
+            os.path.dirname(__file__),
+            "resources_weather",
+            "TRY2010_{:02d}_Jahr.dat".format(self._try_region),
+        )
+        self.weather = dwd_try.read_dwd_weather_file(fn_weather)
```
```suggestion
        if self.file_weather is None:
            self.file_weather = os.path.join(
                os.path.dirname(__file__),
                "resources_weather",
                "TRY2010_{:02d}_Jahr.dat".format(self._try_region),
            )
        self.weather = dwd_try.read_dwd_weather_file(self.file_weather)
```

jnettels reviewed (commented):
I think we should use the default temperature limits if they are not defined by the user.
Also, I added my idea for changing the weather data as a suggestion.

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+            This occurs when ``N_Pers`` or ``N_WE`` are too large.
+
+        """
+        # typtage_combinations = settings["typtage_combinations"]
+        # houses_list = settings["houses_list_VDI"]
```
Is this old code that has been commented out instead of being deleted?

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+            index_col=[0, 1, 2],
+        )
+
+        # if settings.get("zero_summer_heat_demand", None) is not None:
+        #     # Reduze the value of 'F_Heiz_TT' to zero.
+        #     # For modern houses, this eliminates the heat demand in summer
+        #     energy_factors_df.loc[
+        #         (slice(None), slice(None), "F_Heiz_TT"), ("SWX", "SSX")
+        #     ] = 0
```
Is this old code that has been commented out instead of being deleted?

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+                }
+            )
+
+            # df = pd.concat([df_typ, mytime], axis=1).ffill()
+            # df.drop(df.head(1).index, inplace=True)
```
Is this old code that has been commented out instead of being deleted?

p-snft commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,598 @@
+
+            load_curve_house = df_typ.mul(load_profile_df[house["house_type"]])
+
+            # The typical day calculation inherently does not add up to the
+            # desired total energy demand of the full year. Here we fix that:
+            # houses_dict = {h["name"]: h for h in self.houses}
```
Is this old code that has been commented out instead of being deleted?

p-snft reviewed (changes requested):
There seems to be old code that has been commented out instead of being deleted.

p-snft commented:
As this is a draft, I marked it as such.

Side note: I really like the feature. It's very handy to have generation of time series for SH, DHW, and electricity consumption at hand.

uvchik committed:
Update src/demandlib/vdi/regions.py

Co-authored-by: Joris Zimmermann <joris.zimmermann@siz-energieplus.de>

uvchik committed:
Update src/demandlib/vdi/regions.py

Co-authored-by: Joris Zimmermann <joris.zimmermann@siz-energieplus.de>

uvchik commented:
For me it is a bit rude to write your code into my approach without any comment. I already wrote improvements and got dozen of merge conflicts. If I know somebody else is working on the code I would push more often to avoid such problems.

I do have a different approach for the weather file so I think the best way would be to finish this PR with cleaning the code and adding some tests and then start a new PR for the weather data. That makes it easier to track changes and to focus the discussion. One reason for that is that the original approach is connect to the TRY data. I am fine with adding your own weather data but it must be clear that you really should know what you are doing.

Commit:
Fix deprecated warning

Commit:
Make it possible to fetch weather data by knowing the TRY region

Commit:
Make it possible to fetch weather data by knowing the TRY region

p-snft commented on src/demandlib/vdi/dwd_try.py:
```diff
@@ -46,8 +46,14 @@ def find_try_region(longitude, latitude):
+            "TRY2010_{:02d}_Jahr.dat".format(try_region),
+        )
     # The comments in DWD files before the header are not commented out.
     # Thus we have to search for the line with the header information:
```
Carefully reading the description, it is clear that we are talking about Testreferenzjahr files. However, there are a lot of types of DWD files (i.e. recordings from weather stations), maybe we could abbreviate it to TRY files like DWD does.

jnettels committed:
Allow usage of custom weather file paths

jnettels committed:
Allow house name to be other type than a string

While this is not guaranteed to work, other parts of the code
currently work fine if the house names are e.g. integers.

jnettels commented:
Hi there,

it has been quite a while, but I am still depending on this implementation with its awesome speed.
Is there any chance to get it over the finish line and merge it? Is there something I can do to help?

The changes I require to make it work for me personally are added in my fork and affect the ability to load other weather data. Just leaving the link here for reference:
https://github.com/jnettels/demandlib/tree/features/add-vdi-from-lpagg

p-snft commented on examples/vdi_profile_example.py:
```diff
@@ -0,0 +1,103 @@
+
+from demandlib import vdi
+
+# The following dictionary has been created by "workalendar"
```
```suggestion
# The following dictionary can also be created by "workalendar"
```

Maybe this is more clear.

p-snft commented:
@jnettels: Honestly, I'm not really deep into demandlib. I'm willing to review and give feedback. If you have an improved version, you can file a PR against this branch.

jnettels commented:
Thanks for your answer. My "improvements" concerning the weather data were already suggested, once with a rude (still sorry about that!) push to this PR, once with the "suggestion" feature. @uvchik already mentioned that he had a different approach for the weather data, so I do not think creating a new PR for that is helpful now.

jnettels committed:
Fix several deprecation and future warnings from pandas 2.2

uvchik committed:
Merge branch 'dev' of https://github.com/oemof/demandlib into features/add-vdi-from-lpagg

jnettels committed:
Add option 'zero_summer_heat_demand'

Set heat demand on all summer days to zero. This is not part of VDI 4655

jnettels committed:
Add geometry as extra requirements

jnettels committed:
Remove unused files

jnettels committed:
Clean code with flake8 and Black, fix deprecations

jnettels committed:
Improve handling of ImportError

jnettels committed:
Allow usage of fixed season time intervals

jnettels committed:
Allow if-else to be covered by test

House types other than 'EFH' and 'MFH' cannot be properly
tested, because the code would fail before these lines.

jnettels committed:
Add new test file for VDI4655 profile generation

Also add new test dependencie 'geopandas' to pyproject.toml

jnettels committed:
Improve docstrings for vdi script files

jnettels committed:
Add new page for VDI4655 to docs

Also update the index files accordingly

jnettels committed:
Apply black code formatting

uvchik committed:
Fix small issues

uvchik committed:
Merge branch 'dev' of https://github.com/oemof/demandlib into features/add-vdi-from-lpagg

uvchik committed:
Add manifest

uvchik committed:
Fix manifest

uvchik committed:
Fix manifest

uvchik committed:
Fix pytest warnings

uvchik committed:
Fix black issues

uvchik committed:
Fix manifest

uvchik committed:
Remove Python 3.9 and add 3.12 for tests

uvchik committed:
Fix github actions

uvchik committed:
Extract weather from class

uvchik committed:
Add config tests

uvchik committed:
Automatically get weather data if no data is given

uvchik committed:
Fix black issues

uvchik committed:
Merge branch 'features/add-vdi-from-lpagg-2' of https://github.com/oemof/demandlib into features/add-vdi-from-lpagg

uvchik committed:
Add test for custom weather

uvchik committed:
Introduce climate object

uvchik committed:
Fix black issue

uvchik commented:
Implementation of a leap year is still missing but everything else should be fine now. We may want to merge this and start a new PR just for implementing leap years.

@jnettels In my opinion it is not a good idea to create a second PR. I merged both but it cost me some time.

jnettels commented on src/demandlib/vdi/regions.py:
```diff
@@ -0,0 +1,653 @@
+        self.cloud_coverage = cloud_coverage
+        self.energy_factors = energy_factors
+
+    def from_try_data(self, try_region, hoy=8760):
+        if try_region not in list(range(1, 16)):
+            raise ValueError(
+                f">{try_region}< is not a valid number of a DWD TRY region."
+            )
+        fn_weather = os.path.join(
+            os.path.dirname(__file__),
+            "resources_weather",
+            "TRY2010_{:02d}_Jahr.dat".format(try_region),
+        )
+        weather = dwd_try.read_dwd_weather_file(fn_weather)
+        weather = (
+            weather.set_index(
+                pd.date_range(
+                    datetime.datetime(2010, 1, 1, 0),
+                    periods=hoy,
+                    freq="h",
+                )
+            )
+            .resample("D")
+            .mean()
+        )
+        self.temperature = weather["TAMB"]
+
+        weather.loc[weather["CCOVER"] >= 5, "cloud_category"] = "B"
+        weather.loc[weather["CCOVER"] < 5, "cloud_category"] = "H"
+        self.cloud_coverage = weather["cloud_category"]
+        fn_energy_factors = os.path.join(
+            os.path.dirname(__file__),
+            "vdi_data",
+            "VDI_4655_Typtag-Faktoren.csv",
+        )
+        self.energy_factors = pd.read_csv(
+            fn_energy_factors,
+            index_col=[0, 1, 2],
+        ).loc[try_region]
+        return self
```
Allow custom DWD weather data by exposing ``from_file()``.
```suggestion
   def from_try_data(self, try_region, hoy=8760):
        if try_region not in list(range(1, 16)):
            raise ValueError(
                f">{try_region}< is not a valid number of a DWD TRY region."
            )
        fn_weather = os.path.join(
            os.path.dirname(__file__),
            "resources_weather",
            "TRY2010_{:02d}_Jahr.dat".format(try_region),
        )
        self.from_file(fn_weather, try_region, hoy)

        return self

    def from_file(self, fn_weather, try_region, hoy=8760):
        if try_region not in list(range(1, 16)):
            raise ValueError(
                f">{try_region}< is not a valid number of a DWD TRY region."
            )
        weather = dwd_try.read_dwd_weather_file(fn_weather)
        weather = (
            weather.set_index(
                pd.date_range(
                    datetime.datetime(2010, 1, 1, 0),
                    periods=hoy,
                    freq="h",
                )
            )
            .resample("D")
            .mean()
        )
        self.temperature = weather["TAMB"]

        weather.loc[weather["CCOVER"] >= 5, "cloud_category"] = "B"
        weather.loc[weather["CCOVER"] < 5, "cloud_category"] = "H"
        self.cloud_coverage = weather["cloud_category"]
        fn_energy_factors = os.path.join(
            os.path.dirname(__file__),
            "vdi_data",
            "VDI_4655_Typtag-Faktoren.csv",
        )
        self.energy_factors = pd.read_csv(
            fn_energy_factors,
            index_col=[0, 1, 2],
        ).loc[try_region]
        return self
```

jnettels reviewed (commented):
Thanks for your effort!
I wanted to give this feature another chance of merging, but I had to spend my time adding missing code, tests and docs in an efficient way. I needed a clean base to do that. Also after >2.5 years I could not assume there would still be interest in this specific PR. This PR is targeting v0.2dev, which has long been surpassed*, and the branch was out of sync with dev in many regards. So the most efficient way for me seemed to start a fresh branch and cherry-pick all the necessary commits, which worked fine. It produced a clean history, passed all tests and could have just been merged. For me it was the choice between doing it like this, or not contributing at all (because I could not handle all the changes unrelated to the actual new feature). But I do not mind removing my PR, as long as we get the feature merged.

Now for some feedback:

- You removed the ``file_weather`` argument of ``Region()`` and I do not see an obvious way to use custom weather data. We could add a new function ``from_file()`` to ``Climate()``, which ``from_try_data()`` could also use internally.
- The docs need to be updated with usage of the ``Climate()`` class
- My code supported houses where ``Q_Heiz_a``, ``W_a`` or ``Q_TWW_a`` are not defined. I think this is important, because sometimes you may e.g. only want the heating profiles, but no electricity profile. With my proposed changes, those columns are returned as all NaN, which can easily be filtered by the user.
- ``demandlib.ini`` and ``config.py`` are still present. You used them at one point to store the default values for ``summer_temperature_limit`` and ``winter_temperature_limit``. Now it only contains ``test_entry = 5``, which I assume is also unrelated to the VDI feature. Are these files actually used?
- Footnote*: I am very confused. Is ``dev`` the actual dev branch, or ``v0.2dev``? The latter has updates to the BDEW api (``get_scaled_power_profiles()`` instead of ``get_profile()``), which are now also mixed in with this PR, making it even more confusing...

uvchik commented:
Okay, I agree the PR is allready very complex so I see two things we can do:

1. Discuss your issues as we did it last time. This may lead to a better PR but has the risk of a never ending PR (because of the high complexity).
2. Merge this PR and fix the issues in small dedicated PRs. One PR for each topic. e.g. "add `weather_file_parameter`" or "make it possible to get `NaN`". I think the discussion will have a clearer focus and the difficult discussions will not block the simple ones.

I would strongly prefere the second variant. Nothing happend by chance and I have good reasons why I did what I did.

We will merge the PR to `dev`! It is not the `master` branch. So there is still a developer space to fix the issues in small PRs before it will published as a stable version.

jnettels reviewed (approved):
I am fine solving my personal remaining issues as separate PRs.
This PR now covers lots more than the VDI feature, but if you are sure about all of those other changes, then you have got my approvement 👍

uvchik commented:
Because of the difficult merge conflicts it looks more as it is. But this is one more reason to merge it and start with a cleaner structure. Thank you for all you comments and contributions.
